### PR TITLE
Search Bar

### LINKED ISSUE #43: Search Bar
Add SearchBar to home screen.

### PR DESCRIPTION
**Pull request by Naveen-gali**
### What does this PR do?

- Adds TextInput Custom Component
- Adds Label Custom Component
- Adds SearchBar Component to home screen
- Adds Respective Stories.

### Related Issues

- closes #43 

### How to test

- Clone the Repository.
- Switch to feat/searchBar branch.
- Do yarn to install the dependencies.
- Do yarn run build-stage-ios for ios or yarn run build-stage-android for android.
- You Can see the SearchBar on Screen as shown in below images attached

### Screenshots

|Android ScreenShot           |  Ios ScreenShot|
|:-------------------------:|:-------------------------:|
Screenshot_1683258710 | Simulator Screen Shot - iPhone 14 - 2023-05-05 at 09 17 22 |

### Checklist

- [x] I have addressed all the requirements.
- [x] I have tested this code in both IOS and Android.
- [x] I have checked the code for formatting, syntax, and other best practices.
- [x] I have ensured that my changes do not break any existing functionality.
- [x] I have attached the necessary ScreenShots or Screen Recordings.
- [x] I have self reviewed the PR.

### Packages Added (if any)
| Package | Reason |
| ------- | ------ |
|lodash.debounce| Lodash Package for Attaining Debounce        |

### CODE REVIEW

File: android/app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -8,7 +8,7 @@
       android:icon="@mipmap/ic_launcher"
       android:roundIcon="@mipmap/ic_launcher_round"
       android:allowBackup="false"
-      android:theme="@style/BootTheme">
+      android:theme="@style/BootTheme" android:usesCleartextTraffic="true">
```
Comment (prajnab-mm) [inline android/app/src/main/AndroidManifest.xml]: Why do we need this?

File: package.json [hunk 2]
```diff
@@ -36,7 +38,8 @@
     "react-native-safe-area-context": "^4.5.0",
     "react-native-screens": "^3.20.0",
     "react-native-svg": "^13.8.0",
-    "react-native-vector-icons": "^9.2.0"
+    "react-native-vector-icons": "^9.2.0",
+    "yarn": "^1.22.19"
```
Comment (prajnab-mm) [inline package.json]: Do we need this?

File: src/components/TextInput/TextInput.tsx [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+          },
+          getStyle(),
+          focused && mode !== 'border-less'
+            ? [styles.focused, {borderBottomColor: colors.border}]
+            : null,
+          error ? styles.error : null,
+          getDisabledStyle(),
+          inputStyle,
+        ]}
+        onFocus={() => {
```
Comment (prajnab-mm) [inline src/components/TextInput/TextInput.tsx]: Avoid inline functions

File: src/components/TextInput/TextInput.tsx [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+  errorMessage: {
+    color: Palette.error,
+    marginTop: verticalScale(5),
+    marginLeft: horizontalScale(5),
+  },
+  outline: {
+    borderWidth: 1,
+  },
+  outlineDisable: {
+    // TODO: This Horizon or vertical
```
Comment (prajnab-mm) [inline src/components/TextInput/TextInput.tsx]: Is this addressed?

File: src/screens/HomeScreen/HomeScreen.tsx [hunk 5]
```diff
@@ -48,23 +62,59 @@ const HomeScreen = observer((_props: HomeScreenProps) => {
...
+  rootContainer: {
+    height: '100%',
+  },
   loaderView: {
     height: DeviceUtils.getDeviceWindowHeight,
     justifyContent: 'center',
     alignItems: 'center',
   },
+  scrollViewContentStyle: {
+    // TODO: Ask about scaling.
```
Comment (prajnab-mm) [inline src/screens/HomeScreen/HomeScreen.tsx]: No we don't scale. This is like percentage

File: src/screens/HomeScreen/HomeScreen.tsx [hunk 6]
```diff
@@ -48,23 +62,59 @@ const HomeScreen = observer((_props: HomeScreenProps) => {
...
+    bottom: verticalScale(-25),
+    shadowColor: '#000',
+    shadowOffset: {
+      width: 0,
+      height: verticalScale(4),
+    },
+    shadowOpacity: verticalScale(0.3),
+    shadowRadius: verticalScale(4.65),
+    // TODO: Ask about elevation in hori or verti scale
+    elevation: 8,
```
Comment (prajnab-mm) [inline src/screens/HomeScreen/HomeScreen.tsx:141]: We dont scale this as well

File: src/screens/HomeScreen/HomeScreen.tsx [hunk 7]
```diff
@@ -48,23 +62,59 @@ const HomeScreen = observer((_props: HomeScreenProps) => {
...
+    flexGrow: 1,
+  },
+  searchbar: {
+    position: 'absolute',
+    marginHorizontal: horizontalScale(17),
+    padding: horizontalScale(10),
+    width: horizontalScale(360),
+    borderRadius: horizontalScale(6),
+    bottom: verticalScale(-25),
+    shadowColor: '#000',
```
Comment (prajnab-mm) [inline src/screens/HomeScreen/HomeScreen.tsx]: No hardcode color

File: src/components/TextInput/TextInput.tsx [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+        style={[
+          animation,
+          styles.animatedLabel,
+          mode === 'outline'
+            ? {
+                backgroundColor: colors.background,
+              }
+            : null,
+        ]}>
+        {label ? (
```
Comment (prajnab-mm) [inline src/components/TextInput/TextInput.tsx]: Why is the check here and not at the top before animated view?

File: src/screens/HomeScreen/HomeScreen.tsx (hunk 6 above)
Comment (Naveen-gali) [reply src/screens/HomeScreen/HomeScreen.tsx:141]: As Discussed i have removed the TODO

File: package.json (hunk 2 above)
Comment (Naveen-gali) [reply package.json]: It got added when yarn is done. I removed it as it is not needed.

File: .storybook/storybook.requires.js [hunk 9]
```diff
@@ -6,49 +6,51 @@ import {
   addParameters,
   addArgsEnhancer,
   clearDecorators,
-} from '@storybook/react-native';
+} from "@storybook/react-native";
```
Comment (prajnab-mm) [inline .storybook/storybook.requires.js]: Check why quotes are changing from single to double

File: package.json [hunk 10]
```diff
@@ -12,13 +12,20 @@
...
+    "prettier-format-storybook": "prettier .storybook --write",
+    "prepare": "husky install"
   },
   "dependencies": {
     "@react-native-async-storage/async-storage": "^1.18.1",
     "@react-navigation/native": "^6.1.6",
     "@react-navigation/native-stack": "^6.9.12",
     "axios": "^1.3.4",
+    "lodash.debounce": "^4.0.8",
+    "lodash.throttle": "^4.1.1",
```
Comment (prajnab-mm) [inline package.json]: Why is this here?

File: src/components/Label/Label.tsx [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+  return (
+    <Text style={[styles.label, style]} {...restProps}>
+      {label}
+    </Text>
+  );
+};
+
+const styles = StyleSheet.create({
+  label: {
+    fontSize: verticalScale(15),
```
Comment (prajnab-mm) [inline src/components/Label/Label.tsx]: Why is textStyles not used here? Also Color for the text?

File: src/components/SearchBar/SearchBar.tsx [hunk 12]
```diff
@@ -0,0 +1,34 @@
+import React, {useCallback} from 'react';
+import {StyleSheet} from 'react-native';
+import {TextInput} from '../TextInput';
+import {horizontalScale, verticalScale} from '../../utils';
+import {SearchBarProps} from './SearchBar.types';
+import {CustomIcon, CustomIconNames} from '../CustomIcon';
+import debounce from 'lodash.debounce';
```
Comment (prajnab-mm) [inline src/components/SearchBar/SearchBar.tsx]: Keep all absolute imports together

File: src/components/SearchBar/SearchBar.tsx [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+      autoCorrect={false}
+      left={<CustomIcon name={CustomIconNames.Search} style={styles.icon} />}
+      {...restProps}
+    />
+  );
+};
+
+const styles = StyleSheet.create({
+  icon: {
+    fontSize: verticalScale(20),
```
Comment (prajnab-mm) [inline src/components/SearchBar/SearchBar.tsx]: Use themes and text styles here

File: src/components/SearchBar/SearchBar.tsx [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    alignSelf: 'center',
+    marginHorizontal: horizontalScale(2),
+    paddingVertical: 10,
```
Comment (prajnab-mm) [inline src/components/SearchBar/SearchBar.tsx]: Scale

File: src/components/TextInput/TextInput.tsx [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+    left,
+    right,
+    value,
+    ...restProps
+  } = props;
+  const [focused, setFocused] = useState(false);
+  const {colors} = useThemeColor();
+
+  const getStyle = useMemo(() => {
+    if (mode === 'default') {
```
Comment (prajnab-mm) [inline src/components/TextInput/TextInput.tsx]: Why add this check ? We can just return at the end if mode was not outline or border

File: src/components/TextInput/TextInput.tsx [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+      return;
+    } else if (mode === 'outline') {
+      return styles.outline;
+    } else if (mode === 'border-less') {
+      return styles.borderLess;
+    }
+  }, [mode]);
+
+  const getDisabledStyle = useMemo(() => {
+    if (mode === 'outline' && editable === false) {
```
Comment (prajnab-mm) [inline src/components/TextInput/TextInput.tsx]: Here you should add a common check called editable and if true then return. Also, we can just do if(editable) or if(!editable)

File: src/components/TextInput/TextInput.tsx [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+  const renderInputField = () => {
+    return (
+      <InputField
+        style={TextInputStyles}
+        onFocus={onFocus}
+        onBlur={onBlur}
+        placeholder={placeholder}
+        editable={editable}
+        onChangeText={onChangeText}
+        {...restProps}
```
Comment (prajnab-mm) [inline src/components/TextInput/TextInput.tsx:215]: What happens when i pass these above props from parent in restProps

File: package.json (hunk 10 above)
Comment (Naveen-gali) [reply package.json]: I have removed it as it is not needed.

File: src/components/TextInput/TextInput.tsx (hunk 17 above)
Comment (Naveen-gali) [reply src/components/TextInput/TextInput.tsx:215]: I have added the restProps.onFocus to onFocus that is given before and same with onBlur aswell.

Comment (Naveen-gali) [reply src/components/TextInput/TextInput.tsx:215]: const onFocus = () => {
    onFocusHandler();
    setFocused(true);
    restProps.onFocus;
  };

  const onBlur = (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
    onBlurHandler(e.nativeEvent.text);
    setFocused(false);
    restProps.onBlur;
  };

File: src/components/SearchBar/SearchBar.tsx (hunk 14 above)
Comment (Naveen-gali) [reply src/components/SearchBar/SearchBar.tsx]: Fixed

File: src/components/SearchBar/SearchBar.tsx (hunk 13 above)
Comment (Naveen-gali) [reply src/components/SearchBar/SearchBar.tsx]: Fixed

File: src/components/SearchBar/SearchBar.tsx (hunk 12 above)
Comment (Naveen-gali) [reply src/components/SearchBar/SearchBar.tsx]: Organized Imports

File: src/components/Label/Label.tsx (hunk 11 above)
Comment (Naveen-gali) [reply src/components/Label/Label.tsx]: Added The Color and also FontStyles

File: src/components/TextInput/TextInput.tsx (hunk 16 above)
Comment (Naveen-gali) [reply src/components/TextInput/TextInput.tsx]: Revamped.

File: src/components/Label/Label.tsx [hunk 18]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import {StyleSheet, Text} from 'react-native';
+import {fontStyles} from '../../constants';
+import {useThemeColor} from '../../hooks';
+import {LabelProps} from './Label.types';
+
+export const Label = (props: LabelProps) => {
```
Comment (prajnab-mm) [inline src/components/Label/Label.tsx:7]: Every component should have stories. Please add for this

File: src/components/TextInput/TextInput.tsx [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+        </Animated.View>
+      );
+    }
+    return null;
+  };
+
+  const onFocus = () => {
+    onFocusHandler();
+    setFocused(true);
+    restProps.onFocus;
```
Comment (prajnab-mm) [inline src/components/TextInput/TextInput.tsx]: Will this work? have you tested it?

File: src/components/TextInput/TextInput.tsx [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+
+  const moveLabelDown = () => {
+    Animated.timing(floatingLabel, {
+      toValue: 0,
+      duration: 200,
+      useNativeDriver: true,
+    }).start();
+  };
+
+  const onFocusHandler = () => {
```
Comment (prajnab-mm) [inline src/components/TextInput/TextInput.tsx]: Why is onFocusHandler and onBlurHandler created if its used only in onFocus and onBlur functions?

File: src/components/TextInput/TextInput.tsx [hunk 21]
```diff
@@ -0,0 +1,266 @@
...
+      animation,
+      styles.animatedLabel,
+      mode === 'outline'
+        ? {
+            backgroundColor: colors.background,
+          }
+        : null,
+    ];
+
+    const renderLabelStyle = [
```
Comment (prajnab-mm) [inline src/components/TextInput/TextInput.tsx]: It would be better to keep animatedLabelStyle and renderLabelStyle outside renderAnimatedLabel and use memo

File: src/components/Label/Label.tsx (hunk 18 above)
Comment (Naveen-gali) [reply src/components/Label/Label.tsx:7]: Added.

File: src/components/TextInput/TextInput.tsx (hunk 19 above)
Comment (Naveen-gali) [reply src/components/TextInput/TextInput.tsx]: Yeah it is working as expected. I tried with alerts and also console logs and they correctly showed up

File: src/components/TextInput/TextInput.tsx [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+      styles.animatedLabel,
+      mode === 'outline'
+        ? {
+            backgroundColor: colors.background,
+          }
+        : null,
+    ];
+  }, [animation, colors.background, mode]);
+
+  const renderLabelStyle = useMemo(() => {
```
Comment (prajnab-mm) [inline src/components/TextInput/TextInput.tsx]: render prefix to be added only for jsx

File: src/components/TextInput/TextInput.tsx [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+    colors.background,
+    colors.disabled,
+    colors.text,
+    editable,
+    focused,
+    labelStyle,
+    mode,
+  ]);
+
+  const renderAnimatedLabel = () => {
```
Comment (prajnab-mm) [inline src/components/TextInput/TextInput.tsx:195]: As discussed, club all similar variables together

File: src/components/TextInput/TextInput.tsx (hunk 19 above)
Comment (prajnab-mm) [reply src/components/TextInput/TextInput.tsx]: You are not even executing the function. How is it working?

File: src/screens/HomeScreen/HomeScreen.tsx [hunk 24]
```diff
@@ -45,22 +47,47 @@ const HomeScreen = observer((_props: HomeScreenProps) => {
 
   const {colors} = useThemeColor();
 
+  const renderSearchBar = () => {
+    return (
+      <View>
```
Comment (prajnab-mm) [inline src/screens/HomeScreen/HomeScreen.tsx]: Why wrapper View here?

File: src/theme/colors.ts [hunk 25]
```diff
@@ -24,6 +24,9 @@ const LightTheme = {
     activeDots: Palette.neutral200,
     passiveDots: Palette.neutral500,
     loaderBackground: Palette.neutral300,
+    placeholder: Palette.neutral700,
+    searchbarShadow: Palette.neutral900,
```
Comment (prajnab-mm) [inline src/theme/colors.ts]: maybe a generic shadow color?

File: src/components/TextInput/TextInput.tsx (hunk 22 above)
Comment (Naveen-gali) [reply src/components/TextInput/TextInput.tsx]: removed render prefix from the variable names.

File: src/theme/colors.ts (hunk 25 above)
Comment (Naveen-gali) [reply src/theme/colors.ts]: renamed to shadow

File: src/screens/HomeScreen/HomeScreen.tsx (hunk 24 above)
Comment (Naveen-gali) [reply src/screens/HomeScreen/HomeScreen.tsx]: removed as discussed in call.

File: src/components/TextInput/TextInput.tsx (hunk 23 above)
Comment (Naveen-gali) [reply src/components/TextInput/TextInput.tsx:195]: fixed.

File: src/screens/HomeScreen/HomeScreen.tsx [hunk 26]
```diff
@@ -83,18 +105,45 @@ const HomeScreen = observer((_props: HomeScreenProps) => {
...
+    borderRadius: horizontalScale(6),
+    marginTop: verticalScale(-25),
+    shadowOffset: {
+      width: 0,
+      height: verticalScale(4),
+    },
+    shadowOpacity: verticalScale(0.3),
+    shadowRadius: verticalScale(4.65),
+    elevation: 8,
+    zIndex: 2,
```
Comment (prajnab-mm) [inline src/screens/HomeScreen/HomeScreen.tsx]: Do we need this?

File: src/components/TextInput/TextInput.tsx [hunk 27]
```diff
@@ -0,0 +1,282 @@
...
+      return (
+        <Animated.View style={animatedLabelStyle}>
+          <Label style={showLabelStyle} label={label} />
+        </Animated.View>
+      );
+    }
+    return null;
+  };
+
+  const inputField = () => {
```
Comment (prajnab-mm) [inline src/components/TextInput/TextInput.tsx]: prefix all JSX functions with render

File: src/screens/HomeScreen/HomeScreen.tsx (hunk 26 above)
Comment (Naveen-gali) [reply src/screens/HomeScreen/HomeScreen.tsx]: No, removed.

File: src/components/TextInput/TextInput.tsx [hunk 28]
```diff
@@ -0,0 +1,282 @@
...
+    }).start();
+  };
+
+  const onFocusHandler = (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
+    moveLabelTop();
+    setFocused(true);
+    onFocus?.(e);
+  };
+
+  const onEndEditingHandler = (
```
Comment (prajnab-mm) [inline src/components/TextInput/TextInput.tsx:125]: What if i do not press on the submit button? I can dismiss keyboard by tapping outside as well. Will the label go back then?

File: src/components/TextInput/TextInput.tsx [hunk 29]
```diff
@@ -0,0 +1,282 @@
...
+    colors.background,
+    colors.disabled,
+    colors.text,
+    editable,
+    focused,
+    labelStyle,
+    mode,
+  ]);
+
+  const animatedLabel = () => {
```
Comment (prajnab-mm) [inline src/components/TextInput/TextInput.tsx]: render prefix

File: src/components/TextInput/TextInput.tsx [hunk 30]
```diff
@@ -0,0 +1,282 @@
...
+        placeholder={placeholder}
+        editable={editable}
+        onChangeText={onChangeText}
+        onEndEditing={onEndEditingHandler}
+        {...restProps}
+      />
+    );
+  };
+
+  const hintAndError = () => {
```
Comment (prajnab-mm) [inline src/components/TextInput/TextInput.tsx]: +1

File: src/components/TextInput/TextInput.tsx (hunk 22 above)
Comment (prajnab-mm) [reply src/components/TextInput/TextInput.tsx]: You removed for all?

File: src/components/TextInput/TextInput.tsx (hunk 30 above)
Comment (Naveen-gali) [reply src/components/TextInput/TextInput.tsx]: added render prefix for all the jsx elements.

File: src/components/TextInput/TextInput.tsx (hunk 29 above)
Comment (Naveen-gali) [reply src/components/TextInput/TextInput.tsx]: added.

File: src/components/TextInput/TextInput.tsx (hunk 27 above)
Comment (Naveen-gali) [reply src/components/TextInput/TextInput.tsx]: added

File: src/components/TextInput/TextInput.tsx (hunk 28 above)
Comment (Naveen-gali) [reply src/components/TextInput/TextInput.tsx:125]: Yes it is working fine in that case aswell.

File: src/App.tsx [hunk 31]
```diff
@@ -32,7 +32,7 @@ function App(): JSX.Element {
     }
   };
 
-  return renderContent();
+  return content();
```
Comment (prajnab-mm) [inline src/App.tsx]: Why was render prefix removed here?

File: src/components/SearchBar/SearchBar.stories.tsx [hunk 32]
```diff
@@ -0,0 +1,52 @@
+import {StoryObj} from '@storybook/react-native';
+import {Decorator} from '../../../.storybook/decorators';
```
Comment (prajnab-mm) [inline src/components/SearchBar/SearchBar.stories.tsx]: Can we add decorators to index file in storybook and import from there?

Comment (Naveen-gali) [reply src/components/SearchBar/SearchBar.stories.tsx]: I tried exporting it though index.ts file and it is throwing me error. so i reverted it.

Comment (Naveen-gali) [reply src/components/SearchBar/SearchBar.stories.tsx]: Prajna, I have resolved the import issue of the decorator. But i am getting this cyclic dependency warning. Apart from that it is working fine as expected.

`Require cycle: .storybook/index.ts -> .storybook/storybook.requires.js -> src/screens/HomeScreen/components/SearchCarousel/SearchCarousel.stories.tsx -> .storybook/index.ts`.

**Review by prajnab-mm — APPROVED** [review]
